specs/ssh_configuration.adoc

```diff
@@ -0,0 +1,149 @@
+= SSH configuration for provisioning cluster
+
+SSH public-key configuration is used to authenticate cluster nodes with
+provisioning node.
+
+== Problem description
+
+The cluster provisioning process needs some communication between the provisioning
+node and cluster nodes. Before that the provisioning node needs some authentication
+to enable the secure communication with cluster nodes. SSH public-key configuration
+is the process of authenticating cluster nodes with provisioning node.
+
+== Use Cases
+
+Needs SSH configuration utility in common to support SSH configuration for
+provisioning cluster.
+
+== Proposed change
+
+SSH public-key configuration have four major process:
+
+1. Generate a key-pair for a user. The method can take the user name and
+   key-pair name as optional parameters, whose defaults are root and tendrl
+   respectively.
+2. Install an SSH key into a user's authorized_keys. SSH public key should be
+   a mandatory parameter, while the user name should be optional. User name
+   should default to root.
+3. Check if the SSH daemon is running and get it's port number.
+4. Check if the SSH incoming port is open.
+
+== Alternatives
+
+None
+
+== Data model impact
+
+None
+
+== Impacted modules
+
+=== Tendrl/common impact
+
+* Create a new ansible module in commons utils to create a new user with
+  SSH-key. If the user already present then it should create only SSH-key for that
+  user.
+```
+Example arguments for the ansible user module is:
+Create a 2048-bit SSH key for user jsmith in ~jsmith/.ssh/id_rsa
+- user:
+    name: jsmith
+    generate_ssh_key: yes
+    ssh_key_bits: 2048
+    ssh_key_file: .ssh/id_rsa
+```
+* Create a new ansible module commons utils to add SSH authorized key
+```
+Example arguments for the ansible authorized_key module is:
+- name: Set authorized key took from file
+  authorized_key:
+    user: charlie
+    state: present
+    key: "{{ lookup('file', '/home/charlie/.ssh/id_rsa.pub') }}"
+```
+* Create a new a function is utils to find ssh daemon status and port number
+```
+- Port number is identifiable using cat /var/run/sshd.pid, run the command
+  cat /var/run/sshd.pid using ansible to get current process id of sshd.
+- If the pid is not present when ssh daemon is running. If pid is not present then
+  return result like ssh daemon is not running.
+- Python code for find port number using pid
+    import psutil
+    p = psutil.Process(pid)
+    p.name() # to print service name
+    p.connections()
+  sample output:
+      [pconn(fd=3, family=10, type=1, laddr=('::1', 22), raddr=('::1', 54960), status='LISTEN')]
+- Parse the output and return the port number
+- Checkout the status in output if is listen then return result like port is open
```

Comment:
Err? Listening on the port isn't the same as the port being open. Port being open means that the firewall configuration is proper. However, that's not in the scope of this specification, so for now, just report that the service is running and listening on a specific port.